virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -702,3 +691,113 @@ def calculate_soil_evaporation(
     return DataArray(  # TODO check this
         (evaporative_flux / flux_to_mm_conversion) / timestep_conversion_factor
     )
+
+
+def find_lowest_neighbour(neighbours: list, elevation: np.ndarray) -> list:
+    """Find lowest neighbour for each grid cell from digital elevation model.
+
+    This function finds the cell IDs of the lowest neighbour for each grid cell. This
+    can be used to determine in which direction surface runoff flows.
+
+    Args:
+        neighbours: list of neighbour IDs
+        elevation: elevation, [m]
+
+    Returns:
+        list of lowest neighbour IDs
+    """
+    lowest_neighbour = []
+    for cell_id, neighbors_id in enumerate(neighbours):
+        downstream_id_loc = np.argmax(elevation[cell_id] - elevation[neighbors_id])
+        lowest_neighbour.append(neighbors_id[downstream_id_loc])
+
+    return lowest_neighbour
+
+
+def find_upstream_cells(lowest_neighbour: list) -> list:
+    """Find all upstream cell IDs for all grid cells.
+
+    This function identifies all cell IDs that are upstream of each grid cell. This can
+    be used to calculate the water flow that goes though a grid cell.
+
+    Args:
+        lowest_neighbour: list of lowest neighbour cell_ids
+
+    Returns:
+        list of all upstream IDs for each grid cell
+    """
+    upstream_ids: list = [[] for i in range(len(lowest_neighbour))]
+
+    for down_s, up_s in enumerate(lowest_neighbour):
+        upstream_ids[up_s].append(down_s)
+
+    return upstream_ids
+
+
+def accumulate_surface_runoff(
+    drainage_map: dict,
+    surface_runoff: np.ndarray,
+    accumulated_runoff: np.ndarray,
+) -> np.ndarray:
+    """Calculate accumulated surface runoff for each grid cell.
+
+    This function takes the accumulated surface runoff from the previous timestep and
+    adds all surface runoff of the current time step from upstream cell IDs.
+
+    The function currently raises a `ValueError` if accumulated runoff is negative.
+
+    Args:
+        drainage_map: dict of all upstream IDs for each grid cell
+        surface_runoff: surface runoff of the current time step, [mm]
+        accumulated_runoff: accumulated surface runoff from previous time step, [mm]
+
+    Returns:
+        accumulated surface runoff, [mm]
+    """
+
+    upstream_id_list = [[i for i in drainage_map[x]] for x in drainage_map.keys()]
+
+    for cell_id, upstream_id in enumerate(upstream_id_list):
+        accumulated_runoff[cell_id] += np.sum(surface_runoff[upstream_id])
```

Comment:
I think this could be:

```suggestion
    upstream_ids = (range(drainage_map[x]) for x in drainage_map.keys())

    for cell_id, upstream_id in enumerate(upstream_ids):
        accumulated_runoff[cell_id] += np.sum(surface_runoff[upstream_id])
```

Then you avoid creating extra lists unnecessarily.

Reply:
@vgro Is this the suggestion that isn't working for you? Can I ask why?

Reply:
yes, @alexdewar , I get the following error message : TypeError: 'list' object cannot be interpreted as an integer

I also noticed that when I tried to make the typing more specific, for example `list[int]`, I get a lot of mypy error messages. I think you pointed out something very important here. I will look into it again with a fresh brain next week and see how I can tidy this up. Thank you :-)

Reply:
Ahh ok. I think I misread this code. You're just trying to get the *values* out of `drainage_map` and `enumerate` over them, right?

In that case, this might work:

```suggestion
    for cell_id, upstream_ids in enumerate(drainage_map.values()):
        accumulated_runoff[cell_id] += np.sum(surface_runoff[upstream_ids])
```

But yes, more precise typing will help with figuring this out.

Reply:
This solution is very neat and works well. I updated the types and I hope that mypy will pick up arising problems now